### DISCUSSION TITLE
Is it possible to pack dll files and *.dna to xll?

### DISCUSSION THREAD

**Discussion by gagmeng** [augustoproiete/exceldna-unpack#1] [Q&A] [answered]
Is it possible to pack dll files and *.dna to xll?

Thanks.

**Comment by augustoproiete** [comment]
If you are building an Excel Add-In with Excel-DNA using a class library with Visual Studio, then the `.dll` files as well as the `.dna` files will be automatically packed into the `.xll` during the build after you install the [`ExcelDna.AddIn` nuget package](https://www.nuget.org/packages/ExcelDna.AddIn/).

If you're not using Visual Studio, then you can manually do the packing by using the [`ExcelDnaPack.exe`](https://github.com/Excel-DNA/ExcelDna/tree/master/Distribution).

**Comment by gagmeng** [comment]
I tried to unpack a xll file with ExcelDnaUnpack, and then used ExcelDnaPack.exe to pack the unpacked file into xll file. An error occurred while running EXCEL with error message of "The file format and extension of "*.xll" do not match. The file may be damaged or unsafe. Do not open unless you trust its source. Do you still want to open it?".
Thanks for your quick reply.

Best regards,
gagmeng

**Comment by augustoproiete** [comment]
@gagmeng Could you show the list of files that were unpacked from your `.xll` and describe in detail what steps did you take before you ran `ExcelDnaPack.exe`?

There are a number of things you'd have to do and consider. From the top of my head:

* You'll have to make manual updates to the unpacked `.dna` file to remove the `packed:` prefix from the `ExternalLibrary` and `Reference` elements as well as add the `.dll` extension
* You should not pack `EXCELDNA.LOADER.dll` and `EXCELDNA.INTEGRATION.dll` as they are already part of the `.xll`
* To avoid issues, I'd recommend you should use a ["clean" `.xll`](https://github.com/Excel-DNA/ExcelDna/tree/master/Distribution) that you download from the Excel-DNA main repo

**Comment by gagmeng** [comment]
I recorded a video to record my operation process. The video file is as follows：
[https://mega.nz/#!bBhmFAzS!vRuyQtnXaAFqke6geE7a26x1LHt1GZv6AHNph_EJItw](url)

The files used in the video are as follows：
[https://mega.nz/#!2cwhjarA!qOIIpU77QenPKxc-BHs95UBkA6bUjWqRgq8PK4aMj1Y](url)
[https://mega.nz/#!7c41XQqZ!92ibujOQmsrcn6tTavhWA1eIVLEDCEIEqA5wJ0ZI5k8](url)

Thank you for your kind guidance.
Best regards,
gagmeng

**Comment by gagmeng** [comment, marked as answer]
Solve it, Thanks.
